Correct enum indentation in vimscript

JimZhouZZY opened the pull request:
Specially handle commas within an enum block to correct indentation issue in #16289.

And improved test for indentation.

JimZhouZZY committed:
Correct enum indentation issue #16289

JimZhouZZY committed:
Remove a useless conditional statement

JimZhouZZY committed:
Place the special handling of the enum block in a more appropriate position.

zzzyxwvut commented on runtime/autoload/dist/vimindent.vim:
```diff
@@ -461,7 +465,7 @@ export def Expr(lnum = v:lnum): number # {{{2
         endif
     endif
-
+   
```
Trailing whitespace characters.

JimZhouZZY replied:
Removed. Thanks!

zzzyxwvut commented on runtime/autoload/dist/vimindent.vim:
```diff
@@ -588,6 +592,12 @@ export def Expr(lnum = v:lnum): number # {{{2
             return base_ind
         endif
     endif
+ 
```
Ditto.

zzzyxwvut commented on runtime/autoload/dist/vimindent.vim:
```diff
@@ -588,6 +592,12 @@ export def Expr(lnum = v:lnum): number # {{{2
+    if line_A.text =~ STARTS_ENUM
+        line_A->CacheEnumBlock()
+    elseif line_A.lnum->IsInside('EnumBlock')
+        return shiftwidth() 
```
Ditto.

zzzyxwvut commented on runtime/autoload/dist/vimindent.vim:
```diff
@@ -314,6 +314,10 @@ const STARTS_CURLY_BLOCK: string = '\%('
 
+# STARTS_ENUM {{{3
+
+const STARTS_ENUM: string = $'^\s*enum\>\s*'
```
Do you really need string interpolation here?

zzzyxwvut replied:
This pattern should also account for `export`.

zzzyxwvut commented on runtime/autoload/dist/vimindent.vim:
```diff
@@ -839,6 +849,25 @@ def CacheBracketBlock(line_A: dict<any>) # {{{2
+
+    var pos: list<number> = getcurpos()
+    var startlnum: number = line_A.lnum + 1
+    var endlnum: number = search($'^\s*endenum$', 'nW') - 1
```
Do you really need string interpolation here?

zzzyxwvut replied:
Comments can follow `endenum`s, so `'^\C\s*endenum\>'`; else,  
you run the risk of matching some unrelated `endenum`.

JimZhouZZY replied:
I am not very familiar with Vim’s regular expressions. Thank you for the guidance. I've corrected it.

JimZhouZZY committed:
Modified test for enum

JimZhouZZY committed:
Modified test as @zzzyxwvut suggested

> I would also consider adding a more complicated test, e.g.:
>
> ```viml
> " START_INDENT
> enum Digits
> INVALID(v:numbermax),  # The null value.
> ZERO(0 * v:numbermin), ONE(2 - 1),
> TWO(1 + 1), THREE(9 / 3), FOUR(1 * 4),
> FIVE(1 + 2 + 2), SIX(36 / 3 / 2), SEVEN(7), EIGHT(2 * 2 * 2),
> NINE(3 + 3 + 3)
> const value: number
> def new(this.value)
> enddef
> endenum
> " END_INDENT
> ```

zzzyxwvut commented on runtime/autoload/dist/vimindent.vim:
```diff
@@ -588,6 +592,12 @@ export def Expr(lnum = v:lnum): number # {{{2
         endif
     endif
+ 
+    if line_A.text =~ STARTS_ENUM
```
```suggestion
    if line_A.text =~# STARTS_ENUM
```

zzzyxwvut commented on runtime/autoload/dist/vimindent.vim:
```diff
@@ -839,6 +849,25 @@ def CacheBracketBlock(line_A: dict<any>) # {{{2
 enddef
 
+def CacheEnumBlock(line_A: dict<any>) # {{{2
+    if line_A.text !~ '^\s*enum\s'
```
```suggestion
    if line_A.text !~# '^\s*enum\s'
```

JimZhouZZY committed:
Remove trailing whitespace characters.

JimZhouZZY committed:
- Update test to a more complex one.
- Coreect indentation while inputing.

JimZhouZZY committed:
Clean trial blank line

JimZhouZZY committed:
Use buffer vriable instead of global variable

JimZhouZZY committed:
Better solution

JimZhouZZY committed:
Merge branch 'master' into master

JimZhouZZY committed:
Match  and  more precisely

zzzyxwvut commented on runtime/autoload/dist/vimindent.vim:
```diff
@@ -1078,6 +1052,24 @@ def ContinuesBelowBracketBlock( # {{{3
+        cursor(lnum, 1)
+
+        var enum_pos = search('^\C\s*enum\>\s', 'bnW')
+        var endenum_pos = search('^\C\s*endenum\>\s', 'bnW')
```
```suggestion
        var enum_pos = search('^\C\s*\%(export\s\)\=\s*enum\s\+\S\+', 'bnW')
        var endenum_pos = search('^\C\s*endenum\>', 'bnW')
```

The flag combination for `search()` can be tricky; the above  
code seems to follow in `PrevCodeLine()`'s footsteps.  If the  
cursor is inside `EnumBlock`, shouldn't it seek _forward_ for  
`endenum` and _backward_ for `enum`?  Will it work when there is  
more than one `enum` definition in the same file?

JimZhouZZY replied:
For the `export enum` case I think we should also map 'enum' with 'export' modifier. It is previously not covered.

```vimscript
--- a/runtime/autoload/dist/vimindent.vim
+++ b/runtime/autoload/dist/vimindent.vim
@@ -172,6 +172,7 @@ const MODIFIERS: dict<string> = {
     def: ['export', 'static'],
     class: ['export', 'abstract', 'export abstract'],
     interface: ['export'],
+    enum: ['export'],
 }
```

JimZhouZZY replied:
> If the
cursor is inside EnumBlock, shouldn't it seek forward for
endenum and backward for enum? 

No. This logic may fail. Consider a case

```vim script
enum Color
    Red,
    Green,
    Blue
endenom                  #         <-- Missed

var foo: number = 1 #         <-- This line may fail

enum Subject           #         <-- Missed
    Math,
    Physics
endenum
```
In this case, If we seek for `enum` backward and `endenum` forward from the marked line, we might miss the `endenum` right above and the `enum` right below. As a result, the function could incorrectly conclude that the marked line is inside an EnumBlock.

So the simple logic that I came up with is to both seek above. If we find `enum` first, then it must be inside an EnumBlock, else if we find `endenum` first, it should be outside an enum block.

This works unless there's an EnumBlock inside an EnumBlock, as shown below. In this case we might need to handle it using a stack (like what we did for BracketBlocks), but I think there's no(or rare) such syntax on EnumBlocks so far.

```vimscript
enum Color
    enum Subject
        Math,
        Physics
    endenum

    Red,
    Green
endenum
```

JimZhouZZY replied:
I just noticed that I did make a sign mistake in the comparison expression.

I thought the return value when searching backward was relative to the current position so I wrote `<` sign. (However it just worked, unexpectedly... Bruh…)

Corrected. Thank you again for making me rethink.

zzzyxwvut replied:
Fair enough.  It escaped my notice that the cursor is only  
moved once, hence the need to search in one direction.

There is still some room for improvement.

We can do better by avoiding the second lookup as soon as it  
is known that there is no `enum` above.  It is also worth the  
effort to verify that the found `enum` and/or `endenum` line is  
not the current line (remembering about `NonCommentedMatch`'s  
[timeouts](https://github.com/vim/vim/blob/157397edffbc5c375ce54e5e26ee0cabd30a8c73/runtime/autoload/dist/vimindent.vim#L1271)); otherwise, to reject this line for the related  
indentation.  See #16310.

zzzyxwvut commented on runtime/autoload/dist/vimindent.vim:
```diff
@@ -1078,6 +1052,24 @@ def ContinuesBelowBracketBlock( # {{{3
+
+        var enum_pos = search('^\C\s*enum\>\s', 'bnW')
+        var endenum_pos = search('^\C\s*endenum\>\s', 'bnW')
+        
```
Trailing whitespace characters.

JimZhouZZY replied:
Removed for consistency.

zzzyxwvut commented on runtime/indent/testdir/vim.in:
```diff
@@ -956,6 +956,7 @@ FIVE(1 + 2 + 2), SIX(36 / 3 / 2), SEVEN(7), EIGHT(2 * 2 * 2),
 NINE(3 + 3 + 3)
 const value: number
 def new(this.value)
+return this.value
```
```suggestion
def new(value: number)
this.value = value
```
This is a constructor definition (`:help E1365`).

h-east replied:
@zzzyxwvut `def new(this.value)` is a valid constructor syntax. However, neither return nor assignment is required.
See `:h E1390`.

zzzyxwvut replied:
The [suggested test](https://github.com/vim/vim/pull/16293/commits/8737bb883d9486922caccb653ca98494bb28d196) had made use of that syntax.  However,  
with a [newer commit](https://github.com/vim/vim/pull/16293/commits/f0549fe57d693864b2a93a94dce9cf53ca7329e5), additional indentation was introduced  
for testing with a `return this.value` statement; and `return`  
_with a value_ is malformed syntax for constructors.  In order  
to keep that indentation and use valid syntax, the above  
suggestion was offered.

Otherwise, constructors equally support shorthand assignment  
and/or regular parameters, as well as `return`s _with no value_.

For example,

```vim
vim9script

enum Test
    INSTANCE(1260, "twelve hundred and threescore", true)

    const number: number
    const string: string

    def new(this.number, string: string, _)
        if strlen(string) < 24
            this.string = string
        else
            return
        endif
    enddef
endenum
```

zzzyxwvut commented on runtime/indent/testdir/vim.ok:
```diff
@@ -956,6 +956,7 @@ enum Digits
     NINE(3 + 3 + 3)
     const value: number
     def new(this.value)
+	return this.value
```
```suggestion
    def new(value: number)
	this.value = value
```
This is a constructor definition (`:help E1365`).

JimZhouZZY replied:
Modified. Thank you for pointing that out.

h-east replied:
@JimZhouZZY https://github.com/vim/vim/pull/16293#discussion_r1897493893

This code is correct even without defining new() because of the default-constructor mechanism.
See `:h default-constructor`.

Well, since this is just an indentation test code, there is no need to write the code strictly. :smile:

JimZhouZZY committed:
Update runtime/indent/testdir/vim.in

Co-authored-by: Aliaksei Budavei <32549825+zzzyxwvut@users.noreply.github.com>

JimZhouZZY committed:
Update runtime/indent/testdir/vim.ok

JimZhouZZY committed:
Remove trailing whitespace characters

JimZhouZZY committed:
Update combination for search()

Co-authored-by: Aliaksei Budavei <32549825+zzzyxwvut@users.noreply.github.com>

JimZhouZZY committed:
add modifier DYLD_LIBRARY_PATH=/opt/homebrew/opt/python@3.13/Frameworks/Python.framework/Versions/3.13/lib:
HOME=/Users/jimzhou
HOMEBREW_CELLAR=/opt/homebrew/Cellar
HOMEBREW_PREFIX=/opt/homebrew
HOMEBREW_REPOSITORY=/opt/homebrew
INFOPATH=/opt/homebrew/share/info:
LANG=en_US.UTF-8
LOGNAME=jimzhou
OLDPWD=/Users/jimzhou/code/vim
PATH='/opt/homebrew/opt/llvm/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Library/Apple/usr/bin:/Applications/VMware Fusion.app/Contents/Public:/Users/jimzhou/Programs:/opt/homebrew/Cellar/python@3.10:/opt/homebrew/Cellar/python@3.12'
PWD=/Users/jimzhou/code/vim
PYTHONIOENCODING=utf-8
SHELL=/bin/zsh
SHLVL=1
SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.uzqRhgSAmr/Listeners
TERM=xterm-256color
TERM_PROGRAM=Apple_Terminal
TERM_PROGRAM_VERSION=455
TERM_SESSION_ID=BCBB8226-679A-408D-A035-8416CD4775F6
TF_ALIAS=fuck
TF_HISTORY=$'vim --version\nls\nvim e.vim\nvim --clean e.vim\nvim --clean e.vim\nls\ncd ..\ngit clone\nhttps://github.com/JimZhouZZY/vim.git\ngit clont https://github.com/JimZhouZZY/vim.git'
TF_SHELL=zsh
TF_SHELL_ALIASES=$'..=\'cd ..\'\ndir=ls\nf=fuck\nl=ls\nla=\'ls -la\'\nll=\'ls -l\'\nmatlab=/Applications/MATLAB_R2024b.app/bin/matlab\nmcscp=\'scp -i ~/code/minecraft/server_root.pem\'\nmcserver=\'sh ~/code/minecraft/ssh.sh\'\nminecraft=mc\nnethack=\'~/bin/nethack\'\npip=pip3.12\npip3=pip3.12\nproxy=$\'\\n    export http_proxy=socks5://127.0.0.1:7890;\\n    export https_proxy=socks5://127.0.0.1:7890;\\n    export all_proxy=socks5://127.0.0.1:7890;\\n    export no_proxy=socks5://127.0.0.1:7890;\\n    export HTTP_PROXY=socks5://127.0.0.1:7890;\\n    export HTTPS_PROXY=socks5://127.0.0.1:7890;\\n    export ALL_PROXY=socks5://127.0.0.1:7890;\\n    export NO_PROXY=socks5://127.0.0.1:7890;\'\npy=python3.12\npython=python3.12\npython3=python3.12\nrun-help=man\nsl=ls\nsudoit=\'sudo \'\nswap=swapit\nunproxy=$\'\\n    unset http_proxy;\\n    unset https_proxy;\\n    unset all_proxy;\\n    unset no_proxy;\\n    unset HTTP_PROXY;\\n    unset HTTPS_PROXY;\\n    unset ALL_PROXY;\\n    unset NO_PROXY\'\nunrar=unar\nwhich-command=whence'
TMPDIR=/var/folders/qy/qxbg55_x1xlfgs6fstbczdzc0000gn/T/
USER=jimzhou
XPC_FLAGS=0x0
XPC_SERVICE_NAME=0
__CFBundleIdentifier=com.apple.Terminal for

JimZhouZZY committed:
add '\s' back to search endenum

JimZhouZZY committed:
Corect search and compare logic
